Added samples and documentation (#26)
kubeflow/kfserving: docs/README.md

## Specifying a Service
### Resource Definition
This top level resource definition is shared by all Kubernetes Custom Resources.

| Field       |  Value      | Description |
| ----------- | ----------- | ----------- |
| kind       | Service                       | https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#types-kinds |
| apiVersion | serving.kubeflow.org/v1alpha1 | https://kubernetes.io/docs/reference/using-api/api-overview/#api-versioning |
| metadata   | [Metadata](#Metadata)         | https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata |
| spec       | [Spec](#Spec)                 | https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status |

> | status     | [Status](#Status)             | https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status |

Reviewer: Do we need two fields here? seems between label and annotation one can do the job? Or annotation is named more specifically?

Author: These are kubernetes standards. At a high level, labels are for identifying resources, and annotations are used for extensibly tuning implementations without mussing with the spec.

See: https://vsupalov.com/kubernetes-labels-annotations-difference/

### Metadata

> The Metadata section of the resource definition contains resource identification fields and specialized flags. Name and namespace are immutable and if changed, will result in the creation of a new resource alongside the old one. Label and annotations are mutable. Labels and annotations will be applied to all relevant subresources of the Service enabling users to apply the same grouping at the pod level and passthough any necessary annotations.

Reviewer 2: passthough -> pass though?

* Modify the KFService to set the new ModelSpec under the `canary`.
* Set trafficPercent to 0
* Observe the KFService until the new revision progresses
    * Note: traffic is still sent to "default"
* Execute any inspection or testing against `canary`
    * Note: The revision will be available under a specialized URI (e.g. revisionName.name.namespace.cluster.internal)
* Once complete, change the key `canary` to `default` and delete the original `default` section.
* The underlying resource will swap traffic which can be observed in `status`

#### Canary Rollout
For the most advanced deployments, users may follow the Canary Rollout Pattern. It is executed using the following:

* Start with a KFService that has a "stable" or "golden" default definition.
* Modify the KFService to set the new ModelSpec under the `canary`.
* Set trafficPercent to some integer value (e.g. 10)
* Observe the KFService until the new revision progresses from "latestCreatedRevisionName" to "latestReadyRevisionName"
    * Note: `100 - trafficPercent` is still sent to "default"
* Continue incremented trafficPercent in accordance with your desired canarying strategy

> * Once complete, change the key `canary` to `default` and delete the original `default` section.

Author: nice catch!

* The underlying resource will swap traffic which can be observed in `status`

#### Rollback
One of the most compelling characteristics of "infrastructure-as-code" interfaces like the Kubernetes Resource Model is the ability to revert a configuration to a last known golden state. The default/canary pattern defines the entire current state, enabling GitOps rollback.

* revert to previously safe commit
* kubectl apply KFService.yaml

> It's possible to revert to a commit the contains `canary` in the spec. The state will be reconciled to rollback to a configuration that runs both `default` and `canary`. This is possible with the data model, but somewhat of an anti-pattern for safe rollbacks. We recommend reverting to KFService spec that contains only `default`.

Reviewer 3: /the/that/

Author: +1

### Resources
The default implementation will attempt to provide sane defaults for CPU and MEMORY. If more control is needed, we defer to [Specifying Resources in Kubernetes](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container).

By default, a Service will be configured with requests and limits of:
* 1 CPU
* 2 GB MEM

In addition to specifying nvidia/gpu count, users may target specific GPUs on specific platforms by using annotations (e.g. "cloud.google.com/gke-accelerator": "nvidia-tesla-t4"):

> ### Replicas

Reviewer 3: Do we want to document the auto scale down to 0 case which is default?

Author: Yes. I'll add a note on this.

Replicas will be surfaced as part of `status.canary` and `status.default`. It is important to note that `spec.minReplicas` and `spec.maxReplicas` apply to both `default` and `canary`. This may result in large resource usage during rollout if `spec.minReplicas` is set to a large value.

```
status:
  default:
    name: "my-model-endpoint-1"
    replicas: 9
    traffic: 90
  canary 
    name: "my-model-endpoint-2"
    replicas: 1
    traffic: 10
```

### Conditions
Conditions provide realtime feedback to users on the underlying state of their deployment. As the feature set of Service grows, additional Conditions will be added to the set. We will use the convention of positive-true conditions; the resource is healthy if there are no conditions who's status is False. If no conditions are present on the resource (e.g. Ready is not present), the resource is not healthy. Only one condition for each type can be available.  

| Type        | Description |
| ----------- | ----------- | 

> | Ready                | Becomes True when everything is are ready. |

Reviewer 3: /is are ready/is ready/

Author: +1

| RevisionsReady       | Beomces True when the underlying revisions are ready. |
| ContainersHealthy    | Becomes True when containers are successfully pulled, started, and pass readinessChecks. |
| ResourcesProvisioned | Becomes True when minReplicas matches provisionedReplias. If provisioning error detected, reports False immediately. |

### Tensorflow

> TODO Elaborate

Reviewer 4: Will you add it in this PR?

Author: Sure -- we didn't know this before. Thanks for the reminder.

# KFServing

> A KFService is a "unit of model serving". Users may have many KFServices, including different definitions across dev, test, and prod environments, as well as across different model use cases. KFServices are defined using the [Kubernetes Custom Resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) standard. 

Reviewer 5: s/"unit of model serving"/`unit of model serving`

> A KFService is a "unit of model serving". Users may have many KFServices, including different definitions across dev, test, and prod environments, as well as across different model use cases. KFServices are defined using the [Kubernetes Custom Resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) standard. 

Reviewer 5: How about highlight kfserving is based on kative serving? This can help the dev/user understand this project easily.

Author: We're intentionally decoupling Knative from our documentation. It's important that knative is just an implementation detail to maintain flexibility if we end up not being able to use it or want to support other options.

Reviewer: @ellis-bigelow thanks - the issue with decoupling too much is that it becomes an entire new nomenclature to understand, use and work towards. The goal should be to have a very little disruption for knative users, if that`s the primary backing tech here, so that artifacts, ecosystem which emerges around it can be readily reused. 

I gave a talk last Kuebcon around the potential of using knative for ML pipelines, and got great feedback from core kube users because it was explicitly stating the knative can be used, and in what ways. 
https://www.slideshare.net/AnimeshSingh/ai-machine-learning-pipelines-with-knative

Author: I saw your talk!

Feel free to come to one of our KFServing Weekly syncs to bring this up. Just let me know when you'd like to discuss and I'll add it to our agenda. This issue is reasonably contentious, and this is probably the wrong forum to debate it :)

Reviewer: Thanks @ellis-bigelow. 

Do we have one next week - can sync up there. I am seeing this message - 'This event has been canceled and removed from your calendar.'

Author: Every Wednesday. Fridays are special topics and this weeks is cancelled.

## Specifying a Service
### Resource Definition
This top level resource definition is shared by all Kubernetes Custom Resources.

| Field       |  Value      | Description |
| ----------- | ----------- | ----------- |

> | kind       | Service                       | https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#types-kinds |

Reviewer 2: Service -> KFService?

> | kind       | KFService                     | https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#types-kinds |

Reviewer 5: Instead of using links directly here, how about use "[xxx](xxx)"? This may more clear.

Author: Good point. Let me resolve these.

| metadata.labels      | Map<String, String> | A set of key value pairs denoting arbitrary information (e.g integrate with MLOps Lifecycle systems or organize many Models). |
| metadata.annotations | Map<String, String> | A set of key value pairs used to enable specific features available in different kubernetes environments (e.g., Google Cloud ML Engine vs AzureML). |

### Spec
The Spec section of the resource definition encapsulates the desired state of the User's serving resources. Changes made to the spec will enacted upon the underlying servers in an eventually consistent manner. This infrastructure-as-code paradigm enables the coveted GitOps design pattern, where configurations are checked into git and may be reverted and applied to restore previous configurations. All fields are mutable and may be applied idempotently.

| Field       | Value       | Description |
| ----------- | ----------- | ----------- |
| spec.default               | [ModelSpec](#ModelSpec) | The default traffic route serving a ModelSpec. |
| spec.canary                | [ModelSpec](#ModelSpec) | An optional traffic route serving a percent of traffic. |

> | spec.canary.trafficPercent | Integer                 | The amount of traffic to sent to the canary, if specified. |

Reviewer 6: ```suggestion
| spec.canary.trafficPercent | Integer                 | The amount of traffic to sent to the canary (defaults to 0). |
```

#### Blue Green Rollout
The simpliest deployment strategy relies on KFService's default spec. The rollout workflow is identical to a Kubernetes Deployment. Changes to the CRD will create a new set of serving pods. Once the resources are available for serving, traffic will be flipped to the new resources. Users may observe the status of the KFService to gain insight into the current state of the Blue Green rollout.

#### Pinned Rollout
For more cautious and manual workflows, users may follow the Pinned Release pattern. It is executed using the following:

* Start with a KFService that has a "stable" or "golden" default definition.
* Modify the KFService to set the new ModelSpec under the `canary`.

> * Set trafficPercent to 0

Reviewer 6: ```suggestion
* By default canary trafficPercent is set to 0, so no traffic is routed to canary
```

* Observe the KFService until the new revision progresses
    * Note: traffic is still sent to "default"
* Execute any inspection or testing against `canary`

>     * Note: The revision will be available under a specialized URI (e.g. revisionName.name.namespace.cluster.internal)

Reviewer 6: would the canary would also be available in a revision independent URI? e.g. name.namespace.canary.cluster.internal

Author: Yes! I've cleaned this up.

> * Once complete, change the key `canary` to `default` and delete the original `default` section.

Reviewer 6: ```suggestion
* Optionally route some traffic to canary by setting canary.trafficPercent e.g. canary.trafficPercent=10 to route 10% of traffic to the canary revision
* Once complete, change the key `canary` to `default`, delete the trafficPercent attribute and the original `default` section.
```

Reviewer 6: well, duh, reading on I see you already cover this :)

* The underlying resource will swap traffic which can be observed in `status`

#### Canary Rollout
For the most advanced deployments, users may follow the Canary Rollout Pattern. It is executed using the following:

* Start with a KFService that has a "stable" or "golden" default definition.
* Modify the KFService to set the new ModelSpec under the `canary`.
* Set trafficPercent to some integer value (e.g. 10)

> * Observe the KFService until the new revision progresses from "latestCreatedRevisionName" to "latestReadyRevisionName"

Author: Thanks I'm getting rid of these refs. The exact mechanism for exposing these details is TBD but I think we should stay away from Knative "revision" concepts.

>     * Note: `100 - trafficPercent` is still sent to "default"

Author: Tried to clean this up.

* Continue incremented trafficPercent in accordance with your desired canarying strategy
* Once complete, change the key `canary` to `default` and delete the original `default` section and `trafficPercent`.
* The underlying resource will swap traffic which can be observed in `status`

#### Rollback
One of the most compelling characteristics of "infrastructure-as-code" interfaces like the Kubernetes Resource Model is the ability to revert a configuration to a last known golden state. The default/canary pattern defines the entire current state, enabling GitOps rollback.

* revert to previously safe commit
* kubectl apply KFService.yaml

> It's possible to revert to a commit that contains `canary` in the spec. The state will be reconciled to rollback to a configuration that runs both `default` and `canary`. This is possible with the data model, but somewhat of an anti-pattern for safe rollbacks. We recommend reverting to KFService spec that contains only `default`.

Author: I'll remove the harsh language but leave in the recommendation.

Replicas will be surfaced as part of `status.canary` and `status.default`. It is important to note that `spec.minReplicas` and `spec.maxReplicas` apply to both `default` and `canary`. This may result in large resource usage during rollout if `spec.minReplicas` is set to a large value.

```
status:
  default:
    name: "my-model-endpoint-1"
    replicas: 9
    traffic: 90
  canary 
    name: "my-model-endpoint-2"
    replicas: 1
    traffic: 10
```

### Conditions
Conditions provide realtime feedback to users on the underlying state of their deployment. As the feature set of Service grows, additional Conditions will be added to the set. We will use the convention of positive-true conditions; the resource is healthy if there are no conditions who's status is False. If no conditions are present on the resource (e.g. Ready is not present), the resource is not healthy. Only one condition for each type can be available.  

| Type        | Description |
| ----------- | ----------- | 
| Ready                | Becomes True when everything is ready. |

> | RevisionsReady       | Beomces True when the underlying revisions are ready. |

Reviewer 7: Beomces -> Becomes

| ContainersHealthy    | Becomes True when containers are successfully pulled, started, and pass readinessChecks. |
| ResourcesProvisioned | Becomes True when minReplicas matches provisionedReplias. If provisioning error detected, reports False immediately. |

### Tensorflow
### ModelSpec
| Field       | Value       | Description |
| ----------- | ----------- | ----------- |

> | modelUri       | String                                                                                             | URI pointing to Saved Model assets |

Author: I think this is correct as is. The Go standard is URI but most specs just take uri. Open to changing this in the future.
